cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -445,133 +447,182 @@ func (m *McmManager) SetMachineDeploymentSize(machinedeployment *MachineDeployme
-					klog.Infof("Machine %q priority is already set to 1, hence skipping the update", machine.Name)
-					break
+		node, err := m.getNodeForMachineUntilDeadline(nodeName, conflictRetryInterval, retryDeadline)
+		if err != nil && !errors.IsNotFound(err) {
+			return fmt.Errorf("unable to get Node object %s for machine %s, cannot proceed with scale-in of machines, Error: %v", nodeName, machine.Name, err)
+		}
+		if deletetaint.HasToBeDeletedTaint(node) {
+			// Don't update priority annotation if the taint is present
+			continue
+		}
+		clone := machine.DeepCopy()
+		val, ok := clone.Annotations[machinePriorityAnnotation]
+		if ok && val != "3" {
+			clone.Annotations[machinePriorityAnnotation] = "3"
+			for {
+				_, err := m.machineClient.Machines(m.namespace).Update(context.TODO(), clone, metav1.UpdateOptions{})
+				if err != nil && time.Now().Before(retryDeadline) {
+					klog.Warningf("could not reset priority annotation for machine %s; Error: %v, will retry in %s", clone.Name, err, conflictRetryInterval)
+					time.Sleep(conflictRetryInterval)
+				} else if err != nil {
+					return fmt.Errorf("could not reset priority annotation on machine %s; timed out, error: %v", clone.Name, err)
 				}
-				mclone.Annotations[machinePriorityAnnotation] = "1"
-			} else {
-				mclone.Annotations = make(map[string]string)
-				mclone.Annotations[machinePriorityAnnotation] = "1"
 			}
+		}
+	}
+	return nil
+}
 
-			_, err = m.machineClient.Machines(machine.Namespace).Update(context.TODO(), mclone, metav1.UpdateOptions{})
-			if err != nil && time.Now().Before(retryDeadline) {
-				klog.Warningf("Unable to update Machine object %s, Error: %s , will retry in %s", machine.Name, err, conflictRetryInterval)
-				time.Sleep(conflictRetryInterval)
-				continue
-			} else if err != nil {
-				// Timeout occurred
-				klog.Errorf("Unable to update Machine object %s, Error: %s", machine.Name, err)
-				return err
+// updateAnnotationOnMachineUntilDeadline returns error only when updating the annotations on machine has been failing consequently and deadline is crossed
+func (m *McmManager) updateAnnotationOnMachineUntilDeadline(mcName string, key, val string, retryInterval time.Duration, deadline time.Time) error {
+	for {
+		machine, err := m.getMachineUntilDeadline(mcName, retryInterval, deadline)
+		if err != nil {
+			return err
+		}
+		if machine.Annotations != nil {
+			if machine.Annotations[key] == val {
+				klog.Infof("Machine %q priority is already set to 1, hence skipping the update", machine.Name)
+				break
 			}
-
-			// Break out of loop when update succeeds
-			break
+			machine.Annotations[key] = val
+		} else {
+			machine.Annotations = make(map[string]string)
+			machine.Annotations[key] = val
 		}
 
-		klog.Infof("Machine %s of machineDeployment %s marked with priority 1 successfully", machine.Name, md.Name)
+		_, err = m.machineClient.Machines(machine.Namespace).Update(context.TODO(), machine, metav1.UpdateOptions{})
```

Comment:
deadline passed in this method is not going to be honoured because you chose to pass context.TODO to Update method and if that method call blocks then deadline is never honoured.

Reply:
addressed in [03c13fc](https://github.com/gardener/autoscaler/pull/215/commits/03c13fc3b37527d0e3ca8a8571237abdb64ec68f)